### PR TITLE
Graphite: Migrate to React (part 6: Remove AngularJS Query Editor)

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

Follow-up on #37590 and plan described in #36481. This PR removes Graphite AngularJS Query Controller that was providing dependencies to React editor. 

Changes:
* Graphite QueryCtrl and all Graphite-related Angular wrappers are removed
* State is provided with a Context object
* Toggle button added (it was provided out of the box for Angular components but React components have to implement it)
* Tests are now using state/reducers
* Decoupled from SergmentSrv

**Which issue(s) this PR fixes**:

Fixes N/A

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by ifrost** [comment]
Hmm... that `targetFull` might be used for. Closing the PR for the time being to double check it.

**Comment by ifrost** [comment]
> Hmm... that `targetFull` might be used for alerting. Closing the PR for the time being to double check it.

Yes, it is used for alerting with sub-queries. I've added a manual testing scenario to cover this.

**Review by ivanahuckova — COMMENTED** [review]
**From manual testing:**
1. I am not sure if this is bug, but in `Tags>change "dc = asia" to "dc = eu` query is not run right away.  You have to press play button.  If I press run query, the result is incorrect, it runs previous query. When I looked how is this behaving on grafana.play I noticed that we run query right away and that play button disappears after we run it first time. This doesn't happen here. 

2. Similar as above, in `Tags > click on the first tag (with name you entered) and select "-- remove tag --" from the dropdown, one tag should be visible "name = servers.requests.count" and graph should show data for all regions` this is visible only if you run query with play button.

3. `Adding functions >  aggregated graph should be displayed` I have to press play button. If I press run query, the result is incorrect - it is not aggregation.

The play button had to be always pressed with tags.

Other than that, all worked as expected. 

**From code review:** 
All looks good. I have found 1 nit pick that I would be too embarrassed to write down, cause it is such a small unimportant/preference thing. So I am not. But I want to ask about testing. I see that we are testing just 1 component and there are not many unit tests of components. So I was wondering if you plan to add more or if you think those there are sufficient.

**Comment by ifrost** [comment]
> **From manual testing:**
> 
> (...)
> 
> The play button had to be always pressed with tags.
> 

Great catch! It was caused by the fact the unpause updates the state and runs onRunQuery (which also updates the state). Effectively the state should be updated only once (only pause property changed), but it was called twice because synchronizing the time range was incorrect. Should be fixed with c8aba454ccfbd5a74c9794565d98312ebea3b016

> **From code review:**
> All looks good. I have found 1 nit pick that I would be too embarrassed to write down, cause it is such a small unimportant/preference thing. So I am not. But I want to ask about testing. I see that we are testing just 1 component and there are not many unit tests of components. So I was wondering if you plan to add more or if you think those there are sufficient.

Ideally I'd like to turn manual test scenarios into integration tests as follow-up PRs :)

**Review by ivanahuckova — APPROVED** [review]
🚀 Nice job!

**Review by ivanahuckova — APPROVED** [review]

File: public/app/plugins/datasource/graphite/state/context.tsx [hunk 1]
```diff
@@ -1,16 +1,77 @@
...
+  const dispatch = useMemo(() => {
+    return createStore((state) => {
+      setState(state);
+    });
+  }, []);
+
+  // synchronise changes provided in props with editor's state
+  const previousRange = usePrevious(range);
+  useEffect(() => {
+    if (previousRange?.raw !== range?.raw) {
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/plugins/datasource/graphite/state/context.tsx:44]: This value is used to call metric/find api which requires time range (and it should match the time range in dashboards and explore). If object changes and values are the same we don't need to update it (though I didn't see a case where the object would change having same values)

**Review by Elfo404 — APPROVED** [review]
Took it for a ride and seems to work as expected 🚀

**Pull request merged**